src/test/java/io/appium/java_client/service/local/ServerBuilderTest.java

```diff
@@ -1,311 +1,291 @@
+                .hasMoreElements(); ) {
+                InetAddress inetAddress = enumIpAddr.nextElement();
+                if (!inetAddress.isLoopbackAddress()) {
+                    InetAddressValidator validator = InetAddressValidator.getInstance();
+                    testIP = inetAddress.getHostAddress().toString();
+                    if (validator.isValid(testIP)) {
+                        break;
+                    }
+                    testIP = null;
+                }
+            }
+            if (testIP != null) {
+                break;
+            }
+        }
+    }
+
+    @After
+    public void tearDown() throws Exception {
+        ofNullable(service).ifPresent(AppiumDriverLocalService::stop);
+
+        if (stream != null) {
+            stream.close();
+        }
+
+        ofNullable(file).ifPresent(fileArg -> {
+            if (file.exists()) {
+                file.delete();
+            }
+        });
+
+        System.clearProperty(APPIUM_PATH);
+        ofNullable(pathToAppiumNodeInProperties).ifPresent(s -> System.setProperty(APPIUM_PATH, s));
+    }
+
+    @Test public void checkAbilityToStartDefaultService() {
+        service = buildDefaultService();
+        service.start();
+        assertEquals(true, service.isRunning());
+    }
+
+    @Test public void checkAbilityToFindNodeDefinedInProperties() {
+        String definedNode = new File("src/test/java/io/appium/java_client/service/local/appium.js").getAbsolutePath();
+        setProperty(APPIUM_PATH, definedNode);
+        assertThat(new File(new AppiumServiceBuilder().createArgs().get(0)), is(new File(definedNode)));
+    }
+
+    @Test public void checkAbilityToUseNodeDefinedExplicitly() {
+        File node = new File("src/test/java/io/appium/java_client/service/local/appium.js");
+        AppiumServiceBuilder builder = new AppiumServiceBuilder().withAppiumJS(node);
+        assertThat(new File(builder.createArgs().get(0)).getAbsolutePath(), is(node.getAbsolutePath()));
+    }
+
+    @Test public void checkAbilityToStartServiceOnAFreePort() {
+        service = new AppiumServiceBuilder().usingAnyFreePort().build();
+        service.start();
+        assertTrue(service.isRunning());
+    }
+
+    @Test public void checkAbilityToStartServiceUsingNonLocalhostIP() throws Exception {
+        service = new AppiumServiceBuilder().withIPAddress(testIP).build();
+        service.start();
+        assertTrue(service.isRunning());
+    }
+
+    @Test public void checkAbilityToStartServiceUsingFlags() throws Exception {
+        service = new AppiumServiceBuilder()
+            .withArgument(CALLBACK_ADDRESS, testIP)
+            .withArgument(SESSION_OVERRIDE)
+            .withArgument(PRE_LAUNCH)
+            .build();
+        service.start();
+        assertTrue(service.isRunning());
+    }
+
+    @Test public void checkAbilityToStartServiceUsingCapabilities() throws Exception {
+        File appDir = new File("src/test/java/io/appium/java_client");
```

Comment:
there are many common root paths. The common part can be represented as a constant, so then one can use new File(parent, basename) constructor, where parent is a constant